pkg/logs/decoder/line_parser.go

```diff
@@ -66,3 +69,116 @@ func (p *SingleLineParser) process(input *DecodedInput) {
 	}
 	p.lineHandler.Handle(NewMessage(content, status, input.rawDataLen, timestamp))
 }
+
+// MultiLineParser makes sure that chunked lines are properly put together.
+type MultiLineParser struct {
+	buffer       *bytes.Buffer
+	flushTimeout time.Duration
+	inputChan    chan *DecodedInput
+	lineHandler  LineHandler
+	parser       parser.Parser
+	rawDataLen   int
+	lineLimit    int
+	status       string
+	timestamp    string
+}
+
+// NewMultiLineParser returns a new MultiLineHandler.
+func NewMultiLineParser(flushTimeout time.Duration, parser parser.Parser, lineHandler LineHandler, lineLimit int) *MultiLineParser {
+	return &MultiLineParser{
+		inputChan:    make(chan *DecodedInput),
+		buffer:       bytes.NewBuffer(nil),
+		flushTimeout: flushTimeout,
+		lineHandler:  lineHandler,
+		lineLimit:    lineLimit,
+		parser:       parser,
+	}
+}
+
+// Handle forward lines to lineChan to process them.
+func (p *MultiLineParser) Handle(input *DecodedInput) {
+	p.inputChan <- input
+}
+
+// Stop stops the handler.
+func (p *MultiLineParser) Stop() {
+	close(p.inputChan)
+}
+
+// Start starts the handler.
+func (p *MultiLineParser) Start() {
+	p.lineHandler.Start()
+	go p.run()
+}
+
+// run processes new lines from the channel and makes sur the content is properly sent when
+// it stayed for too long in the buffer.
+func (p *MultiLineParser) run() {
+	flushTimer := time.NewTimer(p.flushTimeout)
+	defer func() {
+		flushTimer.Stop()
+		// make sure the content stored in the buffer gets sent,
+		// this can happen when the stop is called in between two timer ticks.
+		p.sendLine()
+		p.lineHandler.Stop()
+	}()
+	for {
+		select {
+		case message, isOpen := <-p.inputChan:
+			if !isOpen {
+				//  inputChan has been closed, no more lines are expected
+				return
+			}
+			// process the new line and restart the timeout
+			if !flushTimer.Stop() {
+				// timer stop doesn't not prevent the timer to tick,
+				// makes sure the event is consumed to avoid sending
+				// just one piece of the content.
+				select {
+				case <-flushTimer.C:
+				default:
```

Comment:
Is the following path possible?
 - `flushTimer.Stop()` return false
 -  There is nothing to read from `flushTimer.C` and so the line 139 is executed.
 - `timer to tick` is send at line 142

Reply:
if `flushTimer.Stop()` is `false` it means the timer is already stopped/expired. We are just trying to read `flushTimer.C` to ensure it's empty (and won't be consumed by the next iteration)  before reseting the timer, if it's already empty there is nothing to do before reseting the timer.